design/hostclaim-multitenancy.md

```diff
@@ -0,0 +1,497 @@
+
+When servers are managed on a separate cluster using the identityRef field in
+the machine template to access the BareMetalHost resources,
+pivot can be performed as usual but the tenant cluster will still need the
+cluster hosting Ironic.
+
+## Design Details
+
+### Implementation Details/Notes/Constraints
+
+#### MetaData transfer between HostClaim and BareMetalHost resources
+
+Most meta-data will be synchronized from the BareMetalHost to the HostClaim.
+
+Two annotations (one for labels, one for annotations) on the HostClaim are
+used to keep the knowledge of which meta-data were imported, so that we can
+delete them when they disappear on the BareMetalHost.
+
+Some specific annotations must be synchronized from the HostClaim to the
+BareMetalHost (reboot, refresh inspection). A dedicated domain must be used.
+
+#### Impact on Metal3Data controller
+
+The Metal3Data controller must target either BareMetalHost or HostClaims
+for some template fields:
+
+* ``fromLabel`` and ``fromAnnotation`` where object is ``baremetalhost``.
+* ``fromHostInterface`` and ``fromAnnotation`` in network definition
+
+The solution is to introduce an intermediate abstract object.
+
+#### Impact on NodeReuse
+
+NodeReuse is implemented in the HostClaim controller with the same algorithm
+as in the original workflow, but the content of the label on the BareMetalHost
+is slightly different.
+
+When a machine template is marked for node reuse, the generated HostClaim
+``nodeReuse`` field contains a string identifying the deployment or the
+control-plane it belongs to (using the algorithm that was used at the time of
+BareMetalHost deletion).
+
+When the HostClaim is deleted, the BareMetalHost is tagged with the node reuse
+label ``infrastructure.cluster.x-k8s.io/node-reuse`` with a value
+joining the namespace of the HostClaim and the value of the ``nodeReuse`` field.
+
+When a new HostClaim is created, if the ``nodeReuse`` field is set, the
+claim will try to bind a BareMetalHost with the label
+``infrastructure.cluster.x-k8s.io/node-reuse`` set to the right value.
+
+### Risks and Mitigations
+
+#### Security Impact of Making BareMetalHost Selection Cluster-wide
+
+The main difference between Metal3 machines and HostClaims it the
+selection process where a HostClaim can be bound to a BareMetalHost
+in another namespace. We must make sure that the binding is expected
+from both the owner of BareMetalHost resources and the HostClaim resource,
+especially when we upgrade the metal3 cluster api provider to a version
+supporting HostClaim.
+
+Choosing between HostClaims and BareMetalHost is done at the level of
+the Metal3Machine controller through a configuration flag. When the HostClaim
+mode is activated, all clusters are deployed with HostClaim resources.
+
+For the server administrator, the solution is to enforce that BareMetalHost
+that can be bound to a HostClaim have the field ``hostClaimNamespaces``
+restricting authorized HostClaims to specific namespaces:
```

Comment:
This should be a new resource (e.g. HostDeployPolicy) for these reasons:

1. To avoid further cluttering BMH
2. Users who can edit policies are not necessarily users who own hardware
3. We may want to make these resources world-viewable so that users know the policies that apply to them